Smart content type middleware

olivierdeckers opened the pull request:
This sets the content type header based on the extension of the request path.

Mime type detection is done by leveraging a JVM mechanism. A default set of mime types is added in META-INF. Users can extend these by adding their own mime.types file as well.

Fixes #238

olivierdeckers committed:
Add mime type detection leveraging the extensible JVM mime.types mechanism

kaushik-barodiya commented on zio-http/src/main/scala/zhttp/http/middleware/HttpMiddleware.scala:
```diff
@@ -363,6 +364,19 @@ object HttpMiddleware {
+   * your applications META-INF folder.
+   */
+  def smartContentType[R, E]: HttpMiddleware[R, E] = {
+    val mimeTypes = new MimetypesFileTypeMap()
```
We are creating a map for each request. Better if we create only once as it can degrade the performance.

olivierdeckers replied:
The map is only created once when the middleware is constructed. I verified this in the HelloWorldWithMiddlewares example

kaushik-barodiya commented on zio-http/src/main/scala/zhttp/http/middleware/HttpMiddleware.scala:
```diff
@@ -363,6 +364,19 @@ object HttpMiddleware {
+    HttpMiddleware.make((_, url, _) => {
+      val contentType = mimeTypes.getContentType(url.path.asString)
+      Patch.addHeader(Header.custom("Content-Type", contentType))
+    })((_, _, S) => S)
```
I found a few problems with the implementation
1. Patch we have added to update the Response. Here you are using Patch as a state. Can we return the contentType as state and patch the response to update Header.
2. For HeaderName you may use HttpHeaderNames class
3. This middleware will update the Header of all the routes. Instead, we should update the header of only the route serving static file.

olivierdeckers replied:
I fixed this

kaushik-barodiya commented on example/src/main/scala/example/HelloWorldWithMiddlewares.scala:
```diff
@@ -14,12 +15,15 @@ import java.util.concurrent.TimeUnit
     // this will return result after 5 seconds, so with 3 seconds timeout it will fail
     case Method.GET -> !! / "long-running" => ZIO.succeed(Response.text("Hello World!")).delay(5 seconds)
-  }
+  } +++ static
```
👍 Really liked the way you used the new Endpoint API to create another app. But for this example do you think it would be extra overkill? Instead, we may just add another route in the existing app?

olivierdeckers replied:
It allowed me to easily test with multiple values for content type, but I agree it might be overkill for this example. I'll change it to a simple route

kaushik-barodiya commented on zio-http/src/test/scala/zhttp/middleware/HttpMiddlewareSpec.scala:
```diff
@@ -135,6 +135,15 @@ object HttpMiddlewareSpec extends DefaultRunnableSpec with HttpAppTestExtensions
+        } +
+          testM("set content type header to application/octet-stream if the extension is not known") {
+            assertM(app(Request(url = URL(!! / "test.abcd"))))(isSome(equalTo("application/octet-stream")))
+          }
```
Can we add one more test where we have ContentType Header already set in Response? In that case, we should preserve the header value

olivierdeckers replied:
I added this

olivierdeckers committed:
Add prefix filtering and some test cases

olivierdeckers commented:
I fixed the build issue, but branch coverage is failing when I run it locally: `Coverage is below minimum [64.29% < 66.00%]: Branch:Total`
I'm a bit confused, because the only part I added is fully covered according to the report?

ashprakasan committed:
Doc : HExit (#707)

kaushik143 commented:
@olivierdeckers Can you resolve the conflicts?

olivierdeckers committed:
Fix merge conflicts

scalafmt

olivierdeckers commented:
> @olivierdeckers Can you resolve the conflicts?

done

olivierdeckers committed:
Merge branch 'main' into smart-content-type

tusharmath commented:
@olivierdeckers I think the static file pr #850 by @ashprakasan should cover this use case more comprehensively.
I don't think we need a middleware for it (unless you think otherwise).

Cheers! Thank you so much for the contribution 🙏
